### ISSUE TITLE
Multi Architecture Support

### ISSUE THREAD

**Issue by mstreatfield** [nerdvegas/rez#154]
We have a need to build packages for multiple cpu architectures.  For example, a library will be built three times - once for AVX2, once for AVX and once for SSE2.  We are looking to model this in Rez and I would appreciate some feedback. 

The three 'variants' of the package can all be built on the same host with different compiler flags (so a distributed build is not required).  

At runtime there is a fallback hierarchy of support available - if a the runtime host has AVX2 available, but the package only has AVX this is ok.  This makes it different to the normal package/variant structure where packages are mutually exclusive.  

Currently, we have two proposals:

1. In the `package.yaml` add a new field.  This will be a list of architectures that package supports.  It would be validated against a master list of architectures from `rezconfig`.  At build time, we have custom CMake macros for `add_library` and `add_executable` which automatically creates targets based on the content of this list.  At runtime, rez makes both the current host's architecture available (`{system.architecture}`) as well as the highest level of support for the current package (`{supported_architecture}`).  This allows the `commands` to then set `LD_LIBRARY_PATH` correctly.

2. model the architectures as normal variants (which aren't mutually exclusive).  This means the build only needs to detect the architecture for the current variant to set the correct compiler flags.  At runtime, Rez adds implicit packages for each supported architecture on the current host machine.  The variant sorting (Fede's pull request) then ensures the correct variant of each package is loaded.

Option 2. is closer to what rez currently supports in terms of variants etc so would require the smallest change to rez.  However this feels susceptible to changes in the resolution algorithm which might not work in our favour.  Option 1 is a larger change to rez but feels a bit more concrete and controllable.

I would be interested in your thoughts.  Currently we are siding with Option 2. (after a long time being set on Options 1.).

**Comment by nerdvegas** [comment]
I would vote for option 2 as well. There are a few relevant points that
affect this.

Firstly - the solver change that affected Fede's variant sorting feature is
going to have to be removed (this is the depth-first iterative solve
approach). It's been in testing at Method for months now, and we've hit a
few rare cases where a non-latest package is incorrectly chosen. I remember
we both thought the problem was related to maximum variant count, but it
turns out it's not. There are several edge cases that can cause the issue.
So it simply has to go - I don't see a way around it.

Having said that - relying on variant sorting to enforce choosing a
particular variant is inherently unsafe. Even without changes to solver
behaviour, the solver has no reason not to reject the variant you actually
want, if that's the way the solve happens to go. But more to the point,
since you're not ensuring variant mutual exclusivity, the resulting
'incorrect' solve *won't fail*. I'm willing to bet there will again be edge
cases where you won't get the variant you want, even though you've
specified the arch package in the request.

But all is not as bad as it seems. For the most part, the variant sorting
will actually work, and in edge cases where it doesn't, there are
workarounds - either adding the package in question directly to the request
as well (just after the arch package), or adding a conflict request to
knock out the undesired variants (ie the other arch's).

As well as this, I've been thinking for some time that it would be useful
to add a feature where a variant of a package can be specifically chosen.
There are cases where we need this. This would involve adding a new request
syntax, something like so:

rez-env foo-5:bah-3

In other words: I request foo-5, but I'm only allowing variants of foo that
overlap with 'bah-5'. This feature would give you the means to ensure that
the solve either uses the variant you require, or does not succeed -
exactly what we need.

Hopefully this helps make your decision, let me know if you have any
questions.

Sorry I've been quiet, all my time is currently taken up on that
contentious and not at all difficult problem of asset management!

A

**Comment by mstreatfield** [comment]
Ok, thanks for the response.

> Firstly - the solver change that affected Fede's variant sorting feature is going to have to be removed 

Ha, we were just thinking about turning that on too!

Do you have an alternative plan to ensure the solve has an upper bound in terms of number of packages loaded?

> rez-env foo-5:bah-3

Yes, I think we have discussed this in other threads.  I quite like the `[#]` notation used to represent the selected variant by index, would using that keep the syntax consistent instead of introducing `:`?

I think we would like to be able to influence the variant across multiple packages, without having to always specify it.  For example: `A -> B -> C` and all three packages have two variants, `D` and `E` (which in our case aren't mutually exclusive).  I only want to request `A` (`rez env A:D`) but ensure the 'same' variant of `B` and `C` is also chosen (so the equivalent of `rez env A:D B:D C:D`).

The variant sorting does this for us now (just) but there are a few times it's broken on edge cases (panic follows).

> I would vote for option 2 as well.

Ok.  I'll probably start mocking something up over the next couple of weeks.  It might require a change to the way implicit packages are injected as I'm not sure I can define them statically (due to the fallback) but I might be able to do it if I model them as dependencies... I'll have a play.

> Sorry I've been quiet

No worries.  Would be good to get a bit of closure on the shell quoting/escaping thread if possible.  I don't think it's a `convert_old_commands` issue, so something that affects rex now.  

Also as part of #55 I might refactor the `Shell` plugins as the API is not clear/well documented for writing new plugins.  Let me know if you have anything particular in mind for how it should look otherwise I'll just make a pull request and see how it goes.

**Comment by nerdvegas** [comment]
Hey just touching base on this issue once more.

"Do you have an alternative plan to ensure the solve has an upper bound in terms of number of packages loaded?"

I have an idea yes. But in the meantime I think we're going to be ok - resources2 introduces many optimisations that speed up solves quite a lot, so I think this will make up for longer solve times due to remove the breadth-first feature.

"Yes, I think we have discussed this in other threads. I quite like the [#] notation used to represent the selected variant by index..."

Actually, we cannot refer to variants by index like this. Since you can now install variants into existing packages, there's no guarantee that the index of a variant in the original package.yaml is the same as the index it ends up with when installed (consider a parallel build - we don't know which variant will get installed first).
Once again this points to the need to support more per-variant data in package definition files, one of which would be a human-readable label that you would be able to use in cases such as this (as well as determining the installation subpath for example).

Another take on this (and I think this might be the best long-term approach) is to assume that we will at some point have a new "features" feature, which would allow the user to specify the package they want not just based on version, but based on arbitrary features that the package(/variant) provides. With this in place, we could have certain "features" that are automatically generated for variants. That would allow us to do things like this:

]$ rez-env foo@requires_python-2.7

In other words, a 'requires_<pkgname>' feature is atomatically created for every package/variant requirement. Features themselves are (label, version) tuples. This same feature would allow us to do this kind of thing:

]$ rez-env foo@variant_lavel=py27

(Here's an example of the same feature, unrelated to variant requirements):
]$ rez-env foo-5+ foo@multithreaded==true

(lucky for us, 'true' is a valid verison number. But you could use '1' instead).
